zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java

```diff
@@ -0,0 +1,114 @@
+  final AtomicReference<List<TopicPartition>> assignedPartitions =
+      new AtomicReference<>(Collections.emptyList());
+
+  KafkaCollectorWorker(Properties kafkaConsumerConfig, String topic, Collector collector,
+      CollectorMetrics metrics) {
+    kafkaConsumer = new KafkaConsumer<>(kafkaConsumerConfig);
+    kafkaConsumer.subscribe(Collections.singleton(topic), new ConsumerRebalanceListener() {
+      @Override public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+        assignedPartitions.set(Collections.emptyList());
+      }
+
+      @Override public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+        assignedPartitions.set(Collections.unmodifiableList(new ArrayList<>(partitions)));
+      }
+    });
+    this.collector = collector;
+    this.metrics = metrics;
+  }
+
+  /**
+   * @return Kafka topic partitions currently assigned to this worker for processing. Returned
+   * list is not modifiable.
+   */
+  public List<TopicPartition> assignedPartitions() {
+    return assignedPartitions.get();
+  }
+
+  @Override
+  public void run() {
+    try {
+      LOG.info("Kafka consumer starting polling loop.");
+      while (true) {
+        final ConsumerRecords<byte[], byte[]> consumerRecords = kafkaConsumer.poll(1000);
+        LOG.debug("Kafka polling returned batch of {} messages.", consumerRecords.count());
+        consumerRecords.forEach((cr) -> {
+          metrics.incrementMessages();
+          final byte[] bytes = cr.value();
+
+          if (bytes.length == 0) {
+            metrics.incrementMessagesDropped();
+          } else {
+            // In TBinaryProtocol encoding, the first byte is the TType, in a range 0-16
+            // .. If the first byte isn't in that range, it isn't a thrift.
+            //
+            // When byte(0) == '[' (91), assume it is a list of json-encoded spans
+            //
+            // When byte(0) <= 16, assume it is a TBinaryProtocol-encoded thrift
+            // .. When serializing a Span (Struct), the first byte will be the type of a field
+            // .. When serializing a List[ThriftSpan], the first byte is the member type, TType.STRUCT(12)
+            // .. As ThriftSpan has no STRUCT fields: so, if the first byte is TType.STRUCT(12), it is a list.
+            if (bytes[0] == '[') {
+              collector.acceptSpans(bytes, Codec.JSON, NOOP);
+            } else {
+              if (bytes[0] == 12 /* TType.STRUCT */) {
+                collector.acceptSpans(bytes, Codec.THRIFT, NOOP);
+              } else {
+                collector.acceptSpans(Collections.singletonList(bytes), Codec.THRIFT, NOOP);
+              }
+            }
+          }
+        });
+      }
+    } finally {
+      LOG.info("Kafka consumer polling loop stopped.");
+      LOG.info("Closing Kafka consumer...");
+      // Clear interrupted status of the thread so the Kafka consumer can at least try to shutdown.
+      Thread.interrupted();
```

Comment:
I went back and forth on this a few times. It seems awkard. But, without it, the Kafka consumers will rarely get closed gracefully, because a thread interrupt is the primary reason the polling loop will be terminated and the kafka consumer will throw an exception if the thread has the interrupted status set.

Closing the consumer isn't strictly necessary, although it can reduce the delay before the consumer group rebalances to account for the consumer being gone.

Your reaction is enough to tip me towards taking it out, since it was a 50/50 decision to begin with. I will do that.